--- a/setup.py
+++ b/setup.py
@@ -39,7 +39,7 @@
 # please look in buildconf.py.example for more information
 PYV8_HOME = os.path.abspath(os.path.dirname(__file__))
 BOOST_HOME = None
-BOOST_MT = is_osx
+BOOST_MT = False
 BOOST_STATIC_LINK = False
 PYTHON_HOME = None
 V8_HOME = None
@@ -581,6 +582,7 @@ class develop(_develop):
 pyv8 = Extension(name="_PyV8",
                  sources=[os.path.join("src", file) for file in source_files],
                  define_macros=macros,
+                 undef_macros=["NDEBUG"] if DEBUG else [],
                  include_dirs=include_dirs,
                  library_dirs=library_dirs,
                  libraries=libraries
